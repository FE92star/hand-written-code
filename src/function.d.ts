interface Function {
  myCall(this: any, ctx: any, ...args: any[]): any

  myBind(this: any, ctx: any, ...args: any[]): any

  myApply(this: any, ctx: any, args?: any[]): any
}
