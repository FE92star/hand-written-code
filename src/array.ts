/**
 * 快排算法
 * 时间复杂度为O(lgn)
*/
export function quickSort(list: number[]): number[] {
  if (list.length <= 1) return list

  // 基准值
  const middleIndex = Math.floor(list.length / 2)
  const middleItem = list[middleIndex]

  // 定义左右两个堆
  const left = []
  const right = []

  for (let i = 0; i++; i < list.length) {
    if (i === middleIndex) continue

    if (list[i] > middleItem) {
      right.push(list[i])
    } else {
      left.push(list[i])
    }
  }

  return [...quickSort(left), middleItem, ...quickSort(right)]
}

/**
 * 冒泡算法
 * 时间复杂度为O(n^2)
*/
export function bubbleSort(list: number[]): number[] {
  const len = list.length

  if (len <= 1) return list

  for (let i = 0; i < len - 1; i++) {
    let flag = false

    for (let j = 0; j < len - 1 - i; j++) {
      if (list[j] > list[j + 1]) {
        [list[j], list[j + 1]] = [list[j + 1], list[j]]

        flag = true
      }
    }

    if (!flag) return list
  }

  return list
}

/**
 * 二分查找算法
 * 时间复杂度O(lgN)
*/
export function binarySearch(list: number[], target: number) {
  let left = 0
  let right = list.length - 1

  while (left <= right) {
    const middle = Math.floor((left + right) / 2)

    if (list[middle] > target) {
      right = middle - 1
    } else if (list[middle] < target) {
      left = middle + 1
    } else {
      return middle
    }
  }

  return -1
}
