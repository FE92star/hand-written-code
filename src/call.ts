Function.prototype.myCall = function (this: any, ctx: any, ...args: any[]) {
  // 1. 如果不是对象，则new一个新对象
  if (typeof ctx !== 'object') ctx = new Object(ctx)

  const fnKey = Symbol()
  // 2. this赋值给绑定的对象
  ctx[fnKey] = this
  // 3. 执行函数
  const result = ctx[fnKey](...args)
  // 4. 删除key
  delete ctx[fnKey]
  // 5. 返回执行结果
  return result
}

Function.prototype.myBind = function (this: any, ctx: any, ...args: any[]) {
  // 1. 如果不是对象，则new一个新对象
  if (typeof ctx !== 'object') ctx = new Object(ctx)
  const _this = this

  return function (...newArgs: any[]) {
    const fnKey = Symbol()
    // 2. this赋值给绑定的对象
    ctx[fnKey] = _this
    // 3. 执行函数
    const result = ctx[fnKey](...args, ...newArgs)
    // 4. 删除key
    delete ctx[fnKey]
    // 5. 返回执行结果
    return result
  }
}

Function.prototype.myApply = function (this: any, ctx: any, args?: any[]) {
  // 1. 如果不是对象，则new一个新对象
  if (typeof ctx !== 'object') ctx = new Object(ctx)

  const fnKey = Symbol()
  // 2. this赋值给绑定的对象
  ctx[fnKey] = this
  // 3. 执行函数
  const result = Array.isArray(args) ? ctx[fnKey](...args) : ctx[fnKey]()
  // 4. 删除key
  delete ctx[fnKey]
  // 5. 返回执行结果
  return result
}
