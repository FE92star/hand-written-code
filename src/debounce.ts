/**
 * 防抖函数-函数到设定的延时时间后一定会执行一次
 * 应用场景——用户输入执行搜索函数
 * @param {Function} fn-目标函数
 * @param {Number} delay-设定执行函数的延时时间(ms)
*/

export function debounce(fn: () => void, delay = 50) {
  let timer: NodeJS.Timeout

  return function (...args: any[]) {
    if (timer) clearTimeout(timer)

    timer = setTimeout(() => {
      fn.apply(this, args)
    }, delay)
  }
}
