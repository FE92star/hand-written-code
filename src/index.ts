export * from './new'
export * from './promise'
export * from './instanceof'
