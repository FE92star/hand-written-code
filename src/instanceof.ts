export function myInstanceof(left: any, right: Function): boolean {
  if (!left) return false

  // 原型链递归查询
  return left.__proto__ === right.prototype || myInstanceof(left.__proto__, right)
}
