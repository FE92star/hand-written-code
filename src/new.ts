/**
 * 实现一个new函数（用于实例化一个构造函数）
 * @param {Class} cls-构造函数
 * @param {Array} args-剩余参数
 * 1. 创建一个新对象，它继承构造函数的原型
 * 2. 执行构造函数，将this指向到新对象上
 * 3. 如果执行的结果是一个对象，返回执行结果，否则返回新对象
 */

export function myNew(cls: Function, ...args: any[]) {
  const newObj = Object.create(cls.prototype)
  const res = cls.apply(newObj, args)

  return typeof res === 'object' ? res : newObj
}
