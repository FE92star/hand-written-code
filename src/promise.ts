type ResolveFn = (value: any) => void
type RejectFn = (reason: any) => void

enum PromiseState {
  PENDING = 'pending',
  FULLFILED = 'fullfiled',
  REJECTED = 'rejected'
}

export class MyPromise {
  status = PromiseState.PENDING
  value = undefined
  reason = undefined
  onResolveCallbacks: any[] = []
  onRejectCallbacks: any[] = []

  constructor(excutor: (resolve: ResolveFn, reject: RejectFn) => void) {
    const resolveFn = (value: any) => {
      if (this.status === PromiseState.PENDING) {
        this.status = PromiseState.FULLFILED
        this.value = value

        this.onResolveCallbacks.forEach((fn: any) => fn())
      }
    }

    const rejectFn = (reason: any) => {
      if (this.status === PromiseState.PENDING) {
        this.status = PromiseState.REJECTED
        this.reason = reason

        this.onRejectCallbacks.forEach((fn: any) => fn())
      }
    }

    try {
      excutor(resolveFn, rejectFn)
    } catch (error: any) {
      throw new Error(error)
    }
  }

  then(onFulfilled: ResolveFn, onRejected: RejectFn) {
    if (this.status === PromiseState.FULLFILED) {
      onFulfilled(this.value)
    }

    if (this.status === PromiseState.REJECTED) {
      onRejected(this.reason)
    }

    if (this.status === PromiseState.PENDING) {
      this.onResolveCallbacks.push(() => {
        onFulfilled(this.value)
      })

      this.onRejectCallbacks.push(() => {
        onRejected(this.reason)
      })
    }
  }
}

// @ts-ignore
Promise._all = (iterObj) => {
  // 1. iterObj 必须是一个可迭代对象, 否则, 无法正常进行则抛出错误
  if (!(typeof iterObj === "object" && iterObj !== null && typeof iterObj[Symbol.iterator] === "function")) {
    throw new TypeError(`${iterObj} is not iterable`);
  }
  iterObj = [...iterObj];
  /*
   *  2. 函数返回值为 `<Promise>` 对象, 当参数 `iterObj` 内所有的 `Promise` 成功, 
   *     该 `<Promise>` 对象成功, 成功数据为所有成功的 `Promise` 结果数组,
   *     有一个不成功, 则该 `<Promise>` 不成功, 失败数据为失败原因字符串
   */
  return new Promise((resolve, reject) => {
    const len = iterObj.length;
    let count = 0;
    if (len === 0) return resolve([]);

    const res = new Array(len);
    iterObj.forEach(async (item: any, index: number) => {
      const newItem = Promise.resolve(item);
      try {
        const result = await newItem;
        res[index] = result;
        if (++count === len) {
          resolve(res)
        }
      } catch (err) {
        reject(err);
      }
    })
  })
}
