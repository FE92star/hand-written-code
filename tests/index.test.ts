import { expect, test } from 'vitest'
import { myNew, MyPromise, myInstanceof, transformTree2List, transformList2Tree, type TreeNode, type ListNode, toString } from '../src'
import '../src/call'

function Person(this: any, name: string, age: number) {
  this.name = name
  this.age = age
}

Person.prototype.say = function () {
  return {
    name: this.name,
    age: this.age,
  }
}

const promiseFn = () => {
  return new MyPromise((resolve, reject) => {
    setTimeout(() => {
      resolve('success')
    }, 1000)
  })
}

function getName(this: any) {
  console.log(this.name)
  return this.name
}

const obj = {
  name: 'Bob'
}

const treeData: TreeNode[] = [
  {
    title: 'parent 1',
    id: '0-0',
    pid: '0',
    children: [
      {
        title: 'parent 1-0',
        id: '0-0-0',
        pid: '0-0',
        children: [
          {
            title: 'leaf',
            id: '0-0-0-0',
            pid: '0-0-0'
          },
          {
            title: 'leaf',
            id: '0-0-0-1',
            pid: '0-0-0'
          },
        ],
      },
      {
        title: 'parent 1-1',
        id: '0-0-1',
        pid: '0-0',
      },
    ],
  },
]

const listData: ListNode[] = [
  {
    "id": "0-0",
    "pid": "0",
    "title": "parent 1",
  },
  {
    "id": "0-0-0",
    "pid": "0-0",
    "title": "parent 1-0",
  },
  {
    "id": "0-0-0-0",
    "pid": "0-0-0",
    "title": "leaf",
  },
  {
    "id": "0-0-0-1",
    "pid": "0-0-0",
    "title": "leaf",
  },
  {
    "id": "0-0-1",
    "pid": "0-0",
    "title": "parent 1-1",
  },
]

test('new and instanceof', () => {
  const person = myNew(Person, 'Bob', 31)

  expect(person.say()).toStrictEqual({ name: 'Bob', age: 31 })

  expect(myInstanceof(person, Person)).toBeTruthy()
})

test('async', async () => {
  const data = await promiseFn()

  expect(data).toBe('success')

  expect(getName.myCall(obj)).toBe('Bob')

  expect(getName.myApply(obj)).toBe('Bob')

  expect(getName.myBind(obj)()).toBe('Bob')
})

test('tree', () => {
  expect(transformTree2List(treeData)).toMatchSnapshot()

  expect(transformList2Tree(listData)).toMatchSnapshot()
})

test('number', () => {
  expect(toString(12345678)).toBe('12,345,678')
})
