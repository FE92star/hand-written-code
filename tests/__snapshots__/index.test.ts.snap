// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tree 1`] = `
[
  {
    "id": "0-0",
    "pid": "0",
    "title": "parent 1",
  },
  {
    "id": "0-0-0",
    "pid": "0-0",
    "title": "parent 1-0",
  },
  {
    "id": "0-0-0-0",
    "pid": "0-0-0",
    "title": "leaf",
  },
  {
    "id": "0-0-0-1",
    "pid": "0-0-0",
    "title": "leaf",
  },
  {
    "id": "0-0-1",
    "pid": "0-0",
    "title": "parent 1-1",
  },
]
`;

exports[`tree 2`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "id": "0-0-0-0",
            "pid": "0-0-0",
            "title": "leaf",
          },
          {
            "children": [],
            "id": "0-0-0-1",
            "pid": "0-0-0",
            "title": "leaf",
          },
        ],
        "id": "0-0-0",
        "pid": "0-0",
        "title": "parent 1-0",
      },
      {
        "children": [],
        "id": "0-0-1",
        "pid": "0-0",
        "title": "parent 1-1",
      },
    ],
    "id": "0-0",
    "pid": "0",
    "title": "parent 1",
  },
]
`;
